deepfence_agent/store/updatemon/src/main.rs

```diff
@@ -0,0 +1,475 @@
+        let src_file = BufReader::new(src_file);
+        ZstdDecoder::new(src_file)
+    };
+    let mut dst_file = fs::File::create(dst_path)
+        .await
+        .context("Could not open the binary destination file")?;
+
+    io::copy(&mut decoder, &mut dst_file)
+        .await
+        .context("Could not decompress the archive")?;
+
+    info!("{} decompressed", name);
+
+    Ok(())
+}
+
+async fn supervisorctl<I, S>(args: I) -> anyhow::Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let _ = Command::new("supervisorctl").args(args).status().await?;
+    Ok(())
+}
+
+async fn supervisorctl_start(service: &str) -> anyhow::Result<()> {
+    supervisorctl(&["start", service])
+        .await
+        .context("Could not start the service")
+}
+
+async fn supervisorctl_stop(service: &str) -> anyhow::Result<()> {
+    supervisorctl(&["stop", service])
+        .await
+        .context("Could not stop the service")
+}
+
+async fn supervisorctl_update() -> anyhow::Result<()> {
+    supervisorctl(&["reread"])
+        .await
+        .context("Could not reread the supervisor config")?;
+    supervisorctl(&["update"])
+        .await
+        .context("Could not reset the programs affected by the supervisord configuration change")
+}
+
+fn is_remote_newer(
+    name: &str,
+    local_artifact: &Artifact,
+    remote_artifact: &Artifact,
+) -> anyhow::Result<bool> {
+    let remote_version = Version::parse(
+        &remote_artifact
+            .version
+            .strip_prefix("v")
+            .unwrap_or(&remote_artifact.version),
+    )?;
+    let local_version = Version::parse(
+        &local_artifact
+            .version
+            .strip_prefix("v")
+            .unwrap_or(&local_artifact.version),
+    )?;
+
+    let res = remote_version > local_version;
+    if res {
+        info!(
+            "Upgrading {} from {} to {}",
+            name, local_version, remote_version
+        );
+    }
+
+    Ok(res)
+}
+
+async fn is_lockfile_free(name: &str, local_artifact: &Artifact) -> anyhow::Result<bool> {
+    if let Some(lockfile) = &local_artifact.lockfile {
+        if lockfile.exists() {
+            let mut f = fs::File::open(lockfile)
+                .await
+                .context("Could not open the lockfile")?;
+            let mut boot_id_lockfile = String::new();
+            f.read_to_string(&mut boot_id_lockfile)
+                .await
+                .context("Could not read the lockfile")?;
+
+            let boot_id_local = boot_id().await?;
+
+            if boot_id_local != boot_id_lockfile {
+                drop(f);
+                fs::remove_file(lockfile)
+                    .await
+                    .context("Could not remove the lockfile")?;
+            } else if is_file_locked(&f, None)? {
+                info!("{} is busy, skipping upgrade for now", name);
+                return Ok(false);
+            }
+        }
+    }
+
+    Ok(true)
```

Comment:
Hum, I am not sure this function is safe - between the time you return `true` and the next steps, someone could acquire the lock, could they not?
We probably need to acquire the lock when it's free to prevent that edge case